relayer/app/state.go

```diff
@@ -0,0 +1,110 @@
+package relayer
+
+import (
+	"encoding/json"
+	"os"
+	"sync"
+
+	"github.com/omni-network/omni/lib/errors"
+
+	"github.com/cometbft/cometbft/libs/tempfile"
+)
+
+// State represents the state of the relayer. It keeps track of the last submitted height for each source chain on each destination chain.
+type State struct {
+	mu       sync.Mutex
+	filePath string
+	cursors  map[uint64]map[uint64]uint64 // destChainID -> srcChainID -> height
+}
+
+func NewEmptyState(filePath string) *State {
+	return &State{
+		filePath: filePath,
+		cursors:  make(map[uint64]map[uint64]uint64),
+	}
+}
+
+// Get returns the current state.
+func (s *State) Get() map[uint64]map[uint64]uint64 {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	// Create a copy of the map to avoid race conditions
+	copyMap := make(map[uint64]map[uint64]uint64)
+	for k, v := range s.cursors {
+		innerMap := make(map[uint64]uint64)
+		for k2, v2 := range v {
+			innerMap[k2] = v2
+		}
+		copyMap[k] = innerMap
+	}
+
+	return copyMap
+}
+
+// GetHeight returns the last submitted height for the given destChainID and srcChainID.
+func (s *State) GetHeight(dstID, srcID uint64) uint64 {
+	srcMap, ok := s.Get()[dstID]
+	if !ok {
+		return 0
+	}
+
+	height, ok := srcMap[srcID]
+	if !ok {
+		return 0
+	}
+
+	return height
```

Comment:
You are not locking srcMap while you are reading from it...

Rather simplify to:
```
mu.lock
defer mu.unlock

return s.cursors[dstID][srcID]
```